common/fixation_entry.go

```diff
@@ -0,0 +1,280 @@
+//
+// Its primary use it to implemented "fixated entries": entries that may change over
+// time, and whose versions must be retained on-chain as long as they are referenced.
+// For examples, an older version of a package is needed as long as the subscription
+// that uses it lives.
+//
+// Once instantiated, FixationStore offers 4 methods:
+//    - SetEntry(index, block, *entry): add a new "block" version of an entry "index".
+//    - GetEntry(index, block, *entry): get a copy (and reference) an version of an entry.
+//    - PutEntry(index, block): drop a reference to a version of an entry.
+//    - RemoveEntry(index): mark an entry as unavailable for new GetEntry() calls.
+//    - GetAllEntryIndex(index): get all the entries indices (without versions).
+//
+// How does it work? The explanation below illustrates how the data is stored, assuming the
+// user is the module "packages":
+//
+// 1. When instantiated, FixationStore gets a `prefix string` - used as a namespace
+// to separate between instances of FixationStore. For instance, module "packages"
+// would use its module name for prefix.
+//
+// 2. Each entry is wrapped by a RawEntry, that holds the index, block, ref-count,
+// and the (marshalled) data of the entry.
+//
+// 3. FixationStore keeps the entry indices with a special prefix; and it keeps the
+// and the RawEntres with a prefix that includes their index (using the block as key).
+// For instance, modules "packages" may have a package named "YourPackage" created at
+// block 110, and it was updated at block 220, and package named "MyPakcage" created
+// at block 150. The store with prefix "packages" will hold the following objects:
+//
+//     prefix: packages_Entry_Index_            key: MyPackage      data: MyPackage
+//     prefix: packages_Entry_Index_            key: YourPackage    data: YourPackage
+//     prefix: packages_Entry_Raw_MyPackage     key: 150            data: RawEntry
+//     prefix: packages_Entry_Raw_YourPackage   key: 220            data: RawEntry
+//     prefix: packages_Entry_Raw_YourPackage   key: 110            data: RawEntry
+//
+// Thus, iterating on the prefix "packages_Entry_Index_" would yield all the package
+// indices. Reverse iterating on the prefix "packages_Entry_Raw_<INDEX>" would yield
+// all the Fixation of the entry named <INDEX> in descending order.
+//
+// 4. FixationStore keeps a reference count of Fixation of entries, and when the
+// count reaches 0 it marks them for deletion. The actual deletion takes place after
+// a fixed number of epochs has passed.
+
+type FixationStore struct {
+	storeKey sdk.StoreKey
+	cdc      codec.BinaryCodec
+	prefix   string
+}
+
+// AppendEntry adds a new entry to the store
+func (fs *FixationStore) AppendEntry(ctx sdk.Context, index string, block uint64, entryData codec.ProtoMarshaler) error {
+	// get the latest entry for this index
+	latestEntry, err := fs.getUnmarshaledEntryForBlock(ctx, index, block, types.DO_NOTHING)
```

Comment:
The refcount of a new entry should be 1 - because it is in use as the current entry. And then when a new entry is added, the refcount of the previous one (if any) should be decreased by one.
The current code works correctly, only because the delete is called at the end so the test there for "STALE_ENTRY_TIME" fails for this new entry just added. It's just too subtle.